src/continuous_agg.c

```diff
@@ -886,21 +892,36 @@ ts_continuous_agg_watermark(PG_FUNCTION_ARGS)
 	Dimension *dim;
 	Datum maxdat;
 	bool max_isnull;
-	int64 watermark;
+	Watermark *watermark = (Watermark *) fcinfo->flinfo->fn_extra;
 	Oid timetype;
 	AclResult aclresult;
 
 	if (PG_ARGISNULL(0))
 		ereport(ERROR,
 				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-				 errmsg("invalid continuous aggregate hypertable")));
+				 errmsg("materialized hypertable cannot be NULL")));
+
+	if (watermark != NULL)
+	{
+		/* The flinfo is probably reset when the input argument changes, but
+		 * the documentation isn't clear on this. Check that the hypertable ID
+		 * matches across repeated calls just to be safe. */
+		if (watermark->hyper_id == hyper_id)
+			PG_RETURN_INT64(watermark->value);
```

Comment:
Last allocation will not be freed here I guess and live till fn_mcxt life time?

Reply:
It only has to free if re-allocated. Last call will free automatically when the memory context is destroyed.